zabbix_smartmontools/__init__.py

```diff
@@ -403,33 +544,104 @@ def parseConfig(path=None):
 
     return d
 
-def main():
-    fail_ifNot_Py3()
 
-    cmd = sys.argv[1]
-    host = '"%s"' % (sys.argv[2])
-    config = parseConfig()
+def readConfig(config):
+    '''Read and display important config values for debug.'''
+    try:
+        f = open(config, 'r')
+        text = f.read()
+        f.close()
+
+        print("  Config's main settings:")
+        server = re.search(r'^(?:\s+)?(Server(?:\s+)?\=(?:\s+)?.+)$', text, re.M)
+        if server:
+            print(server.group(1))
+        else:
+            print("Could not find 'Server' setting in config!")
 
-    configError = None
-    if not 'Disks' in config:
-        scanDisks_Out = scanDisks(config, cmd)   # scan the disks
+        serverActive = re.search(r'^(?:\s+)?(ServerActive(?:\s+)?\=(?:\s+)?.+)$', text, re.M)
+        if serverActive:
+            print(serverActive.group(1))
+        else:
+            print("Could not find 'ServerActive' setting in config!")
 
-        configError = scanDisks_Out[0]   # SCAN_OS_NOCMD, SCAN_OS_ERROR, SCAN_UNKNOWN_ERROR
-        diskList = scanDisks_Out[1]
-    else:
-        diskList = config['Disks']
+    except:
+        print('  Could not process config file:\n' + config)
+    finally:
+        print()
 
-    jsonData, senderData = getAllDisks(config, host, cmd, diskList)
 
-    if configError:
-        senderData.append('%s smartctl.info[ConfigStatus] "%s"' % (host, configError))
-    elif not diskList:
-        senderData.append('%s smartctl.info[ConfigStatus] "NODISKS"' % (host))   # if no disks were found
+def sanitizeStr(s):
+    '''Sanitizes provided string in sequential order.'''
+    stopChars = (
+        ('/dev/', ''), (' -d', ''),
+        ('!', '_'), (',', '_'), ('[', '_'), ('~', '_'), ('  ', '_'),
+        (']', '_'), ('+', '_'), ('/', '_'), ('\\', '_'), ('\'', '_'),
+        ('`', '_'), ('@', '_'), ('#', '_'), ('$', '_'), ('%', '_'),
+        ('^', '_'), ('&', '_'), ('*', '_'), ('(', '_'), (')', '_'),
+        ('{', '_'), ('}', '_'), ('=', '_'), (':', '_'), (';', '_'),
+        ('"', '_'), ('?', '_'), ('<', '_'), ('>', '_'), (' ', '_'),
+    )
+
+    for i, j in stopChars:
+        s = s.replace(i, j)
+
+    s = s.strip()
+
+    return s
+
+
+def scanDisks(config, command):
+    '''Determines available disks. Can be skipped.'''
+    disks = []
+    try:
+        p = subprocess.check_output([config['ctlPath'], '--scan'], universal_newlines=True)   # scan the disks
+        error = ''
+    except OSError as e:
+        p = ''
+
+        if e.args[0] == 2:
+            error = 'SCAN_OS_NOCMD'
+        else:
+            error = 'SCAN_OS_ERROR'
+    except Exception as e:
+        try:   # extra safe
+            p = e.output
+        except:
+            p = ''
+
+        error = 'SCAN_UNKNOWN_ERROR'
+        if command == 'getverb':
+            raise
+
+    for l in p.splitlines():
+        match = re.match(r'/dev/(\S+) [-]d (\S+)', l)
+        if match:
+            disks.append((match[1], match[2]))
```

Comment:
```suggestion
            disks.append((match.group(1), match.group(2)))
```
```
$ time zabbix_get -k smartctl.discovery[get,NON] -s debian9
...
TypeError: '_sre.SRE_Match' object is not subscriptable
```

Reply:
Hm, that suggests a different error.  Probably the regex is wrong, because it ought to match every single line.  Can you please show me the output of `smartctl --scan` on that system?

Reply:
```
# smartctl --scan
/dev/sda -d scsi # /dev/sda, SCSI device
/dev/sdb -d scsi # /dev/sdb, SCSI device
/dev/sdc -d scsi # /dev/sdc, SCSI device
/dev/sdd -d scsi # /dev/sdd, SCSI device
/dev/sde -d scsi # /dev/sde, SCSI device
/dev/sdf -d scsi # /dev/sdf, SCSI device
/dev/sdg -d scsi # /dev/sdg, SCSI device
```

Reply:
Then why doesn't the regex match?  Is there a blank line at the end or something?

Reply:
I can't tell. But it's tested multiple times with [group](https://docs.python.org/3.3/library/re.html#re.match.group) on Linux.

Reply:
What version of Python were you using that gave you that error?

Reply:
Python 3.5.3